// @openzeppelin/hardhat-upgrades/dist/utils/simulate-deploy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.simulateDeployImpl = simulateDeployImpl;
const upgrades_core_1 = require("@openzeppelin/upgrades-core");
const deploy_impl_1 = require("./deploy-impl");
// To import an already deployed contract we want to reuse fetchOrDeploy for its ability to validate
// a deployment and record it in the network file. We are able to do this by "simulating" a deployment:
// for the "deploy" part we pass a function that simply returns the contract to be imported, rather than
// actually deploying something.
async function simulateDeployImpl(hre, ImplFactory, opts, implAddress) {
    const { deployData, simulateDeploy } = await getSimulatedData(hre, ImplFactory, opts, implAddress);
    await (0, upgrades_core_1.fetchOrDeploy)(deployData.version, deployData.provider, simulateDeploy, opts, true);
}
/**
 * Gets data for a simulated deployment of the given contract to the given address.
 */
async function getSimulatedData(hre, ImplFactory, opts, implAddress) {
    const deployData = await (0, deploy_impl_1.getDeployData)(hre, ImplFactory, opts);
    const simulateDeploy = async () => {
        return {
            abi: ImplFactory.interface.format(true),
            layout: deployData.layout,
            address: implAddress,
        };
    };
    return { deployData, simulateDeploy };
}
