// @openzeppelin/defender-sdk-network-client/lib/api/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkClient = void 0;
const defender_sdk_base_client_1 = require("@openzeppelin/defender-sdk-base-client");
const PATH = '/networks';
class NetworkClient extends defender_sdk_base_client_1.BaseApiClient {
    getPoolId() {
        return process.env.DEFENDER_POOL_ID || 'us-west-2_94f3puJWv';
    }
    getPoolClientId() {
        return process.env.DEFENDER_POOL_CLIENT_ID || '40e58hbc7pktmnp9i26hh5nsav';
    }
    getApiUrl() {
        // TODO: update to /monitor when available
        return process.env.DEFENDER_API_URL || 'https://defender-api.openzeppelin.com/v2/';
    }
    async listSupportedNetworks(params) {
        return this.apiCall(async (api) => {
            return await api.get(params && params.networkType ? `${PATH}?type=${params.networkType}` : `${PATH}`);
        });
    }
    async listForkedNetworks() {
        return this.apiCall(async (api) => {
            return await api.get(`${PATH}/fork`);
        });
    }
    async createForkedNetwork(network) {
        return this.apiCall(async (api) => {
            return await api.post(`${PATH}/fork`, { ...network, networkType: 'fork' });
        });
    }
    async deleteForkedNetwork(id) {
        return this.apiCall(async (api) => {
            return await api.delete(`${PATH}/fork/${id}`);
        });
    }
    async getForkedNetwork(id) {
        return this.apiCall(async (api) => {
            return await api.get(`${PATH}/fork/${id}`);
        });
    }
    async updateForkedNetwork(id, network) {
        return this.apiCall(async (api) => {
            return await api.put(`${PATH}/fork/${id}`, { ...network, tenantNetworkId: id });
        });
    }
    async listPrivateNetworks() {
        return this.apiCall(async (api) => {
            return await api.get(`${PATH}/private`);
        });
    }
    async createPrivateNetwork(network) {
        return this.apiCall(async (api) => {
            return await api.post(`${PATH}/private`, { ...network, networkType: 'private' });
        });
    }
    async deletePrivateNetwork(id) {
        return this.apiCall(async (api) => {
            return await api.delete(`${PATH}/private/${id}`);
        });
    }
    async getPrivateNetwork(id) {
        return this.apiCall(async (api) => {
            return await api.get(`${PATH}/private/${id}`);
        });
    }
    async updatePrivateNetwork(id, network) {
        return this.apiCall(async (api) => {
            return await api.put(`${PATH}/private/${id}`, { ...network, tenantNetworkId: id });
        });
    }
}
exports.NetworkClient = NetworkClient;
