// @openzeppelin/hardhat-upgrades/dist/defender/propose-upgrade-with-approval.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeProposeUpgradeWithApproval = makeProposeUpgradeWithApproval;
require("../type-extensions");
const upgrades_core_1 = require("@openzeppelin/upgrades-core");
const utils_1 = require("./utils");
const prepare_upgrade_1 = require("../prepare-upgrade");
const client_1 = require("./client");
function makeProposeUpgradeWithApproval(hre, defenderModule) {
    return async function proposeUpgradeWithApproval(proxyAddress, contractNameOrImplFactory, opts = {}) {
        opts = (0, utils_1.enableDefender)(hre, defenderModule, opts);
        const client = (0, client_1.getDeployClient)(hre);
        const network = await (0, utils_1.getNetwork)(hre);
        if (await (0, upgrades_core_1.isBeaconProxy)(hre.network.provider, proxyAddress)) {
            throw new Error(`Beacon proxy is not currently supported with defender.proposeUpgradeWithApproval()`);
        }
        else {
            // try getting the implementation address so that it will give an error if it's not a transparent/uups proxy
            await (0, upgrades_core_1.getImplementationAddress)(hre.network.provider, proxyAddress);
        }
        let proxyAdmin = undefined;
        if (await (0, upgrades_core_1.isTransparentProxy)(hre.network.provider, proxyAddress)) {
            // use the erc1967 admin address as the proxy admin
            proxyAdmin = await (0, upgrades_core_1.getAdminAddress)(hre.network.provider, proxyAddress);
        }
        const implFactory = typeof contractNameOrImplFactory === 'string'
            ? await hre.ethers.getContractFactory(contractNameOrImplFactory)
            : contractNameOrImplFactory;
        const abi = implFactory.interface.formatJson();
        const deployedImpl = await (0, prepare_upgrade_1.deployImplForUpgrade)(hre, proxyAddress, implFactory, {
            getTxResponse: true,
            ...opts,
        });
        const txResponse = deployedImpl.txResponse;
        const newImplementation = deployedImpl.impl;
        const upgradeProposalResponse = await client.upgradeContract({
            proxyAddress: proxyAddress,
            proxyAdminAddress: proxyAdmin,
            newImplementationABI: abi,
            newImplementationAddress: newImplementation,
            network: network,
            approvalProcessId: opts.approvalProcessId,
        });
        return {
            proposalId: upgradeProposalResponse.proposalId,
            url: upgradeProposalResponse.externalUrl,
            txResponse,
        };
    };
}
